File: web/src/shell/runs.ts
import type { RunCursor, RunState, RunsQuery } from '../gen/bindings'

/* What the runs view is narrowed to. Dates are the date fields' own values. */
export type RunFilters = {
  readonly stepId: string | null
  readonly state: RunState | null
  readonly from: string
  readonly to: string
}

export const STATES: readonly RunState[] = ['running', 'ok', 'failed', 'cancelled', 'lost']

/* The unix second a `YYYY-MM-DD` day starts on, here, `plus` days later.
   Built from the parts rather than parsed: `new Date('2026-09-01')` is UTC
   midnight, which is the previous evening west of Greenwich. */
function dayStart(day: string, plus = 0): number | null {
  const [year, month, date] = day.split('-').map(Number)
  if (!year || !month || !date) return null
  return new Date(year, month - 1, date + plus).getTime() / 1000
}

export function asQuery(projectId: string, filters: RunFilters, after: RunCursor | null): RunsQuery {
  return {
    projectId,
    stepId: filters.stepId,
    state: filters.state,
    since: filters.from ? dayStart(filters.from) : null,
    /* The day in `to` is included, so the range ends where the next day starts. */
    until: filters.to ? dayStart(filters.to, 1) : null,
    after,
  }
}

/** Which empty the runs view is looking at, when it is looking at one. */
export type Empty = 'noChecks' | 'nothingRan' | 'noneMatch'

/**
 * The three empties, which are three different sentences.
 *
 * A board with no lane that runs anything has nothing to say and something to
 * offer — configure a step. A board that has one and has never run it is
 * waiting, not misconfigured. And a filter that matches nothing is neither:
 * the runs are there, behind the filter the person set.
 *
 * Telling them apart matters more than it sounds: "No runs match" on a project
 * that cannot run anything sends somebody looking for a filter that is not the
 * problem.
 */
export function emptyBecause(
  checks: number,
  filters: RunFilters,
  filtered: boolean,
): Empty | null {
  if (!filtered) return null
  if (checks === 0) return 'noChecks'
  return narrowed(filters) ? 'noneMatch' : 'nothingRan'
}

/** Whether anything is narrowing the list. */
export const narrowed = (filters: RunFilters): boolean =>
  filters.stepId !== null || filters.state !== null || filters.from !== '' || filters.to !== ''

File: crates/rpc/src/runs.rs
//! The project's runs, as the runs view pages through them.
//!
//! Apart from `board.rs`, which is at its size ceiling and is the shape of a
//! board; this is history across every card on it.

use serde::{Deserialize, Serialize};
use specta::Type;

use crate::board::{Run, RunState};

/// Where the last page ended: the next one starts after this run.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct RunCursor {
    pub started_at: f64,
    pub id: String,
}

/// What to list. Every filter is optional, and they narrow together.
#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct RunsQuery {
    pub project_id: String,
    /// A lane is filtered by the step it runs.
    pub step_id: Option<String>,
    pub state: Option<RunState>,
    /// Unix seconds, inclusive.
    pub since: Option<f64>,
    /// Unix seconds, exclusive.
    pub until: Option<f64>,
    pub after: Option<RunCursor>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct ProjectRun {
    pub run: Run,
    pub card_id: String,
    pub card_title: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct RunsPage {
    pub runs: Vec<ProjectRun>,
    /// Absent on the last page.
    pub next: Option<RunCursor>,
}
